clippy_lints/src/while_pop_unwrap.rs

```diff
@@ -0,0 +1,122 @@
+use clippy_utils::{diagnostics::span_lint_and_then, match_def_path, paths, source::snippet};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, PatKind, Stmt, StmtKind, UnOp};
+use rustc_lint::LateContext;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{symbol::Ident, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Looks for loops that check for emptiness of a `Vec` in the condition and pop an element
+    /// in the body as a separate operation.
+    ///
+    /// ### Why is this bad?
+    /// Such loops can be written in a more idiomatic way by using a while..let loop and directly
+    /// pattern matching on the return value of `Vec::pop()`.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let mut numbers = vec![1, 2, 3, 4, 5];
+    /// while !numbers.is_empty() {
+    ///     let number = numbers.pop().unwrap();
+    ///     // use `number`
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let mut numbers = vec![1, 2, 3, 4, 5];
+    /// while let Some(number) = numbers.pop() {
+    ///     // use `number`
+    /// }
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub WHILE_POP_UNWRAP,
+    style,
+    "checking for emptiness of a `Vec` in the loop condition and popping an element in the body"
+}
+declare_lint_pass!(WhilePopUnwrap => [WHILE_POP_UNWRAP]);
+
+fn report_lint(cx: &LateContext<'_>, pop_span: Span, ident: Option<Ident>, loop_span: Span, receiver_span: Span) {
+    span_lint_and_then(
+        cx,
+        WHILE_POP_UNWRAP,
+        pop_span,
+        "you seem to be trying to pop elements from a `Vec` in a loop",
+        |diag| {
+            diag.span_suggestion(
+                loop_span,
+                "try",
+                format!(
+                    "while let Some({}) = {}.pop()",
+                    ident.as_ref().map_or("element", Ident::as_str),
+                    snippet(cx, receiver_span, "..")
+                ),
+                Applicability::MaybeIncorrect,
+            )
+            .note("this while loop can be written in a more idiomatic way");
+        },
+    );
+}
+
+fn match_method_call(cx: &LateContext<'_>, expr: &Expr<'_>, method: &[&str]) -> bool {
+    if let ExprKind::MethodCall(..) = expr.kind
+        && let Some(id) = cx.typeck_results().type_dependent_def_id(expr.hir_id)
+        && match_def_path(cx, id, method)
+    {
+        true
+    } else {
+        false
+    }
+}
+
+fn is_vec_pop(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
```

Comment:
This method can probably be inlined, it's only used once, and doesn't have much documentation value.